vsp.go

```diff
@@ -57,65 +61,309 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			go func() {
+				err := vsp.buyTickets(cancelCtx, tipHeader, expiry, balanceToMaintain, passphrase)
+				if err != nil {
+					switch {
+					// silence these errors
+					case errors.Is(err, errors.InsufficientBalance):
+					case errors.Is(err, context.Canceled):
+					case errors.Is(err, context.DeadlineExceeded):
+					default:
+						log.Errorf("Ticket purchasing failed: %v", err)
+					}
+					if errors.Is(err, errors.Passphrase) {
+						fatalMu.Lock()
+						fatal = err
+						fatalMu.Unlock()
+						outerCancel()
+					}
+				}
+			}()
+		}
+	}
+}
+
+// buyTickets purchases one or more tickets depending on the spendable balance of the selected account,
+// the specified minimum balance to maintain and the limit per block.
+func (vsp *VSP) buyTickets(ctx context.Context, tip *wire.BlockHeader, expiry int32, maintain int64, passphrase []byte) error {
+	ctx, task := trace.NewTask(ctx, "ticketbuyer.buy")
+	defer task.End()
+
+	wal := vsp.w
+
+	if len(passphrase) > 0 && wal.IsLocked() {
+		err := wal.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	defer wal.LockWallet()
+
+	// Get the account balance to determine how many tickets to buy
+	bal, err := wal.GetAccountBalance(int32(vsp.purchaseAccount))
+	if err != nil {
+		return err
+	}
+
+	spendable := bal.Spendable
+	if spendable < maintain {
+		log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+		return nil
+	}
+
+	spendable -= maintain
+	sdiff, err := wal.Internal().NextStakeDifficultyAfterHeader(ctx, tip)
+	if err != nil {
+		return err
+	}
+
+	buy := int(dcrutil.Amount(spendable) / sdiff)
+	if buy == 0 {
+		log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+		return nil
+	}
+
+	max := int(wal.Internal().ChainParams().MaxFreshStakePerBlock)
+	if buy > max {
+		buy = max
+	}
+
+	//TODO: limit should be set dynamically
+	// limit := 1
+	// if limit > 0 && buy > limit {
+	// 	buy = limit
+	// }
```

Comment:
Restore this, but the limit should be passed as a parameter and should ALWAYS be 1 per my previous comment.